src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/CSharpCodeActionProvider.cs

```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal abstract class CSharpCodeActionProvider
+    {
+        public abstract Task<RazorCodeAction[]> ProvideAsync(RazorCodeActionContext context, IEnumerable<RazorCodeAction> codeActions, CancellationToken cancellationToken);
```

Comment:
So the gist of how this functions now is it takes in a code action context, a list of pre-existing csharp code actions and then remaps/translates them to be Razor code actions right? I wonder if "provide" is the right terminology for how this works. Also i'd suggest making the return value immutable (`IReadOnlyList<RazorCodeAction>`)

Reply:
> and then remaps/translates them to be Razor code actions right
First ensures that specific provider can handle the code action (thereby filtering out code actions which aren't supported by razor) and then does the remap/translate if the code action is supported.

> I wonder if "provide" is the right terminology for how this works.
I kept it as `provide` to keep consistent with the `RazorCodeActionProvider` which is similar except of course it doesn't take a code action. We could do `ProcessAsync`, but we would have to change the `CSharpCodeActionProvider` name as well then. Open to suggestions

> Also i'd suggest making the return value immutable (IReadOnlyList<RazorCodeAction>)
Done.

Reply:
Can leave as is for now.